### PR TITLE
Add scanner support to accept deposit return vouchers (Apps-1642, Apps-1645)

### PR DESCRIPTION
**Pull request by Fabtron**
Deposit return vouchers can now be scanned. This means their accepted and a matching line item is added to the shopping cart. The violation handling has been integrated to ensure that the new feature behavous correctly. 

This PR does not contain any ui.

APPS-1642 / 1645

### How to test?
Integrate this branch into an app usinng [deposit return vouchers](https://snabble.atlassian.net/wiki/spaces/DEV/pages/834207749/Deposit+Return+Vouchers#Codes-for-Testing). 

Test out at least the following. 

1. Scan the top 3 items from the link and add products. Trigger the checkout once you have a positive price.
2. Scan an item multiple times -> A violation message pops up handling the douplicated.
3. Scan the already redeemed -> A violation message pops up handling the already redeemed.

Don't scan the zero amount or the redemption failed! We cannot handle it atm.

### Definition of Done

- [x] Issue is linked
- [x] All requirements of the issue are fulfilled
- [ ] Changelog is updated
- [x] Documentation is updated
- [x] [Self-Review](https://www.nerdwallet.com/blog/engineering/why-you-should-be-doing-self-reviews/)
- [ ] Review with the Product Owner _(Release-Variante o. Minified Build)_

#### App Tests
- [x] Minified build has been tested _(aka. Release Build)_
- [x] Environments have been taken care of _(Production/Staging)_
- [x] Supported languages have been tested
- [ ] Light-/Dark-Mode has been tested
- [ ] Edge-Cases have been tested
- [ ] Android API Levels have been taken care of _(minSdk?)_

#### Testing
- [ ] Tests have been written _(aka Unit-Tests, Integration-Tests, ...)_

### CODE REVIEW

File: core/src/main/java/io/snabble/sdk/shoppingcart/ShoppingCart.kt [hunk 1]
```diff
@@ -694,6 +717,42 @@ class ShoppingCart(
...
+                        fallbackMessage = violation.message
+                    )
+                )
+            }
+    }
+
+    private fun Item.violates(
+        violations: List<Violation>,
+        type: ViolationType,
+    ) = violations.filter { it.type == type }.any { it.refersTo == id }
```
Comment (cmaier) [inline core/src/main/java/io/snabble/sdk/shoppingcart/ShoppingCart.kt:753]: Since violations are a List, it's being filtered first and then it's checked if there is any w/ the matching predicate.

There are two ways to run the more efficiently:

1. Create a Sequence which will improve the processing in our favour
2. Omit the filtering and make it part of the any predicate

File: core/src/main/java/io/snabble/sdk/shoppingcart/ShoppingCartUpdater.kt [hunk 2]
```diff
@@ -278,6 +285,11 @@ internal class ShoppingCartUpdater(
         else -> false
     }
 
+    private fun containsReturnDepositReturnVouchers(signedCheckoutInfo: SignedCheckoutInfo): Boolean {
+        val (_, lineItems) = deserializedCheckoutInfo(signedCheckoutInfo) ?: return false
+        return lineItems.any { it.type == LineItemType.DEPOSIT_RETURN_VOUCHER }
```
Comment (cmaier) [inline core/src/main/java/io/snabble/sdk/shoppingcart/ShoppingCartUpdater.kt:288]: There's only one return statement necessary:

```kotlin
return deserializedCheckoutInfo(signedCheckoutInfo)
    ?.lineItems
    ?.any { it.type == LineItemType.DEPOSIT_RETURN_VOUCHER }
    ?: false
```

File: ui/src/main/java/io/snabble/sdk/ui/scanner/SelfScanningView.java [hunk 3]
```diff
@@ -225,17 +240,23 @@ public void lookupAndShowProduct(List<ScannedCode> scannedCodes, BarcodeFormat b
                 .resolve();
     }
 
-    private void handleCoupon(List<ScannedCode> scannedCodes, String failureMessage) {
+    private void handleReturnDepositVoucher(List<ScannedCode> scannedCodes) {
+        final kotlin.Pair<CodeTemplate, String> codeTemplateScannedCodePair = SelfScanningExtensionsKt.containsReturnDepositVoucher(project, scannedCodes);
+        if (codeTemplateScannedCodePair != null) {
+            SelfScanningExtensionsKt.insertDepositReturnVoucherItem(shoppingCart, codeTemplateScannedCodePair.getFirst(), codeTemplateScannedCodePair.getSecond());
```
Comment (cmaier) [inline ui/src/main/java/io/snabble/sdk/ui/scanner/SelfScanningView.java:246]: With Java interop annotation _@file:JvmName_ you define a matching name for the Java usages!

**Review by cmaier — CHANGES_REQUESTED** [review]
Some improvements to the functional approaches. 👍
